src/python/pants/engine/process.py

```diff
@@ -470,6 +472,30 @@ def __init__(
         super().__init__(msg)
 
 
+class BashBinary(BinaryPath):
+    """The `bash` binary."""
+
+    DEFAULT_SEARCH_PATH = SearchPath(("/usr/bin", "/bin", "/usr/local/bin"))
+
+
+@rule(desc="Finding the `bash` binary", level=LogLevel.DEBUG)
+async def find_bash(search_path: SearchPath) -> BashBinary:
```

Comment:
This argument doesn't look right to me. We don't want any generic `SearchPath` object for this rule; we want the one scoped for `Bash`. If you wanted to keep this setup, you'd do something like subclass `SearchPath` to have `BashSearchPath`.

Instead, can you remove the argument and have only one rule that goes from `() -> BashBinary`? Hardcode the search paths within the rule. This would mean you can't do `await Get(BashBinary)` anymore though - it'd have to be a param to the rule. That should be fine because we will need to execute this code anytime we need to find binaries.

Reply:
We really don't care what search path someone uses. How do you know they're wrong? We only care about our known use and that does pass in the search path we want.

Reply:
To be more concrete, say I have a rule that wants to privilege brew bash. It would not like our default search path and it would want to get its own unique BashBinary different from the one we bootstrapped for satisfying binary search requests.

Reply:
.... and you're insteads don't form valid rule graphs - cycles.

Reply:
Ah - perhaps you meant you don't mind SearchPath having varying content (addressing the brew example), you just worry this rule would somehow use a SearchPath that came from somewhere other than a `Get(BashBinary, SearchPath(...this one exactly please...))`: I guess that could happen is someday we introduced a nullary rule that produced a global SearchPath? But that seems to indicate a bug in the rule graph / processing if so. At any rate - if I've got the concern right, I think the parametrization of rationale addresses. If not, I still do not get the concern over varying the search path values / order.